### PR TITLE
feat(heatmap): enable brushing on categorical charts

### LINKED ISSUE #1170: Heatmap brushing action for categorical data
**Describe the bug**
Brushing action doesn't render overlay mask for categorical data 

**To Reproduce**
Provide data with categorical data for both axes and perform a brushing action 

**Expected behaviour**
The overlay mask is rendered the same way as for the time series data.

**Screenshots**
May-27-2021 14-34-32

**Version (please complete the following information):**
 - Elastic Charts: `@latest`

**Kibana Cross Issues**
https://github.com/elastic/kibana/pull/97978

### LINKED ISSUE #1171: Brushing action crashes the the browser's window
**Describe the bug**
Brushing over numerical categorical data causing the browser's window to crash 

**To Reproduce**
Provide chart data that contain numbers on the X-axis. You can pass `6236f697-f919-4d1b-a697-01c5cce7b81b` to the `xAccessor` from this sample data.

<details>
  <summary>Check sample data</summary>
  
  ```json
[
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Shoes",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 210,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 333
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Shoes",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 44,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 298
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Shoes",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 165,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 215
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Shoes",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": "__other__",
    "b6dcf411-56af-4df5-b572-17602e2825f5": 89
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Shoes",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 150,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 343
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Shoes",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 165,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 215
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Shoes",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 110,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 205
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Shoes",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": "__other__",
    "b6dcf411-56af-4df5-b572-17602e2825f5": 83.33333333333333
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Accessories",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 210,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 333
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Accessories",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 155,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 317
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Accessories",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 110,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 193
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Accessories",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": "__other__",
    "b6dcf411-56af-4df5-b572-17602e2825f5": 65
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Clothing",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 210,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 277.5
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Clothing",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 165,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 248
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Clothing",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 200,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 225
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Men's Clothing",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": "__other__",
    "b6dcf411-56af-4df5-b572-17602e2825f5": 64.125
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Accessories",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 150,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 343
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Accessories",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 200,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 225
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Accessories",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 115,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 144
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Accessories",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": "__other__",
    "b6dcf411-56af-4df5-b572-17602e2825f5": 61.46875
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Clothing",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 150,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 343
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Clothing",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 165,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 215
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Clothing",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": 80,
    "b6dcf411-56af-4df5-b572-17602e2825f5": 210.5
  },
  {
    "79b13663-d722-4554-8746-b35c4a266673": "Women's Clothing",
    "6236f697-f919-4d1b-a697-01c5cce7b81b": "__other__",
    "b6dcf411-56af-4df5-b572-17602e2825f5": 57.96527777777778
  }
]
  ```
</details>

**Expected behaviour**
Brushing actions doesn't crush a browser's window

**Screenshots**
May-27-2021 14-57-51_

**Version (please complete the following information):**
 - OS: macOS
 - Browser: Chrome
 - Elastic Charts: `@latest`

**Kibana Cross Issues**
https://github.com/elastic/kibana/pull/97978

### PR DESCRIPTION
**Pull request by markov00**
## Summary
The brush tool is now available on heatmap with categorical axes. The `onBrushEnd` callback will receive an array of values for both the x and the y coordinates corresponding to the brushed area.

## Details

The existing implementation doesn't handle brushing on heatmap with non-time x axis.
The code was written to cover only that case.
The crash in #1171 was due to a infinite loop caused by a mix of numerical/string values used in example provided.
This was fixed applying a logic for time series and a different one for the categorical one

### To Dos
- [x] cleanup code
- [x] add interaction tests
- [ ] ~remove `onBrushEnd` from the `config` and using the settings one (different PR?)~ https://github.com/elastic/elastic-charts/issues/1214
- [ ] ~fix `brushTool` that goes outside the projection area~ https://github.com/elastic/elastic-charts/issues/1216
- [ ] ~fix the mouse projection area to avoid tooltip on cells that are not visible due to the pagination~ https://github.com/elastic/elastic-charts/issues/1215

## Issues

fix #1170 fix #1171 

### Checklist

- [x] The proper chart type label was added (e.g. :xy, :partition) if the PR involves a specific chart type
- [x] The proper feature label was added (e.g. :interactions, :axis) if the PR involves a specific chart feature
- [x] Whenever possible, please check if the closing issue is connected to a running GH project
- [x] Any consumer-facing exports were added to `packages/charts/src/index.ts` (and stories only import from `../src` except for test data & storybook)
- [ ] This was checked for cross-browser compatibility
- [x] Proper documentation or storybook story was added for features that require explanation or tutorials
- [x] Unit tests were updated or added to match the most common scenarios

### CODE REVIEW

File: packages/charts/src/chart_types/heatmap/layout/viewmodel/viewmodel.ts [hunk 1]
```diff
@@ -358,12 +355,11 @@ export function shapeViewModel(
...
-
     return {
-      x: xStart,
-      y: yStart,
-      width,
-      height: totalHeight,
+      x: Math.round(xStart),
+      y: Math.round(yStart),
+      width: Math.round(width),
+      height: Math.round(totalHeight),
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/heatmap/layout/viewmodel/viewmodel.ts:362]: What is the purpose of rounding this? Would this be a breaking change?

**Review by nickofthyme — APPROVED** [review]
LGTM, tested locally and tested brushing on several stories including the heatmap story.

File: packages/charts/src/chart_types/heatmap/layout/viewmodel/viewmodel.ts (hunk 1 above)
Comment (markov00) [reply packages/charts/src/chart_types/heatmap/layout/viewmodel/viewmodel.ts]: You are right, I've recovered the default implementation here 8468de3
I don't exactly know why, but rounding the values made the brush rectangle appear. I've tried now without rounding and it still works, not sure what was the exact issue during the implementation

**Pull request merged**